Products: Only use one bottom sheet for both AI and manual product type creation

hafizrahman opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of: #11361
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR is part one to do the improvement listed in #11361.

It deals with displaying the list of product type creation options immediately when "Add manually" is selected in the product creation AI bottom sheet, instead of opening to another sheet.

The design is not yet identical to the video in #11361, it shall be done separately. Specifically, these are not added yet:
- Use new heading "Create Product / select product type")
- Split list of product type creation options into sub-lists (Standard, Subscription, Others).

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

1. In a site that's eligible for AI support, go to Products tab,
2. Press "+",
3. On the bottom sheet that appear, tap "Add manually",
4. Ensure the product types creation options are shown.
5. Try an option and ensure it works properly.
6. Also check that analytics are tracked properly for each option selection: `add_product_product_type_selected`

## Video


https://github.com/woocommerce/woocommerce-ios/assets/266376/5258423c-8292-4fe5-8e5f-8ee84d3ff74c



---
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

hafizrahman committed:
Add SwiftUI file for displaying manual product type creation options.

hafizrahman committed:
Display manual product type options content in the SwiftUI view.

hafizrahman committed:
Make ManualProductTypeOptions appear in the AIActionSheet.

hafizrahman committed:
Generate the command inside Coordinator instead.

hafizrahman committed:
Add subtitle (and a quick whitespace fix)

hafizrahman committed:
Pass callback for when a product type creation option is selected.

hafizrahman commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/BottomSheetListSelector/ProductTypeBottomSheetListSelectorCommand.swift:
```diff
@@ -3,7 +3,8 @@ import Yosemite
 /// This includes the remote `ProductType`, whether the product type is virtual, and strings/images shown in the bottom sheet.
 ///
-public enum BottomSheetProductType: Hashable {
+public enum BottomSheetProductType: Hashable, Identifiable {
```
Added `Identifiable` so it can be used inside `ForEach`.

hafizrahman committed:
Prevent possible id duplicate.

hafizrahman committed:
Add comment for clarity.

dangermattic commented:
<!--
  0 Errors
  2 Warnings: This PR is larger than 500 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/183">17.2</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductWithAI/EntryPoint/AddProductWithAIActionSheet.swift:
```diff
@@ -71,20 +79,32 @@ struct AddProductWithAIActionSheet: View {
-                .onTapGesture {
-                    onManualOption()
+
+                if isShowingManualOptions {
```
Nit: you can use `else` here.

hafizrahman replied:
Right! added in https://github.com/woocommerce/woocommerce-ios/pull/11859/commits/b84fd11df840cb27816b1af3ddafb61645348115

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/Add Product/ManualProductTypeOptions.swift:
```diff
@@ -0,0 +1,46 @@
+    var body: some View {
+        ForEach(command.data) { model in
+            HStack(alignment: .top, spacing: Constants.margin) {
+                Image(uiImage: model.actionSheetImage)
+                    .font(.title3)
+                    .foregroundColor(.secondary)
```
The foreground color is not applied to some images. Please ensure the images are rendered as templates for the foreground color to take effect.

Also, we should switch to `foregroundStyle` since `foregroundColor` is deprecating soon.

<img src="https://github.com/woocommerce/woocommerce-ios/assets/5533851/79f814a2-bf88-4f17-bb81-b7465762cc99" width=320 />

hafizrahman replied:
TIL about rendered as template. Updated in https://github.com/woocommerce/woocommerce-ios/pull/11859/commits/f2a78fd450e07e82f76b3e7d1a88ebc236a8ac7e

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/Add Product/ManualProductTypeOptions.swift:
```diff
@@ -0,0 +1,46 @@
+/// View to show the manual product type creation options.
+///
+struct ManualProductTypeOptions: View {
+    private let command: ProductTypeBottomSheetListSelectorCommand
```
Nit: we should send the `BottomSheetProductType` list here since it's the data that we need. `ProductTypeBottomSheetListSelectorCommand` type is associated with the `BottomSheetListSelectorViewController`, which is irrelevant to this view.

```suggestion
    private let items: [BottomSheetProductType]
```

itsmeichigo replied:
It would be even better if `BottomSheetProductType` is moved to a separate file now that it's used in two places.

hafizrahman replied:
Thanks, great suggestion here. I refactored both the parameter and file separation in https://github.com/woocommerce/woocommerce-ios/pull/11859/commits/bc86823cd2e479aae81f486ed0ed660acf198879

itsmeichigo reviewed (commented):
This looks nice! I left some suggestions for improvements below.

hafizrahman committed:
Simplify if-else statement.

hafizrahman committed:
Refactor to use [BottomSheetProductType] for `ManualProductTypeOptions`.

This also moves BottomSheetProductType into its own file (as it's now shared in different places). Localized strings are also updated.

hafizrahman committed:
Update icon styling.

itsmeichigo reviewed (approved):
Everything looks good now! 🚢
